domain/domain.go

```diff
@@ -92,78 +92,75 @@ type Domain struct {
-// It returns the latest schema version, the changed table IDs, whether it's a full load and an error.
-func (do *Domain) loadInfoSchema(handle *infoschema.Handle, usedSchemaVersion int64,
-	startTS uint64) (neededSchemaVersion int64, change *tikv.RelatedSchemaChange, fullLoad bool, err error) {
+// loadInfoSchema loads infoschema at startTS.
+// It returns:
+// 1. the needed infoschema
+// 2. cache hit indicator
+// 3. currentSchemaVersion(before loading)
+// 4. the changed table IDs if it is not full load
+// 5. an error if any
+func (do *Domain) loadInfoSchema(startTS uint64) (infoschema.InfoSchema, bool, int64, *tikv.RelatedSchemaChange, error) {
 	snapshot := do.store.GetSnapshot(kv.NewVersion(startTS))
 	m := meta.NewSnapshotMeta(snapshot)
-	neededSchemaVersion, err = m.GetSchemaVersion()
+	neededSchemaVersion, err := m.GetSchemaVersion()
 	if err != nil {
-		return 0, nil, fullLoad, err
-	}
-	if usedSchemaVersion != 0 && usedSchemaVersion == neededSchemaVersion {
-		return neededSchemaVersion, nil, fullLoad, nil
+		return nil, false, 0, nil, err
 	}
 
-	// Update self schema version to etcd.
-	defer func() {
-		// There are two possibilities for not updating the self schema version to etcd.
-		// 1. Failed to loading schema information.
-		// 2. When users use history read feature, the neededSchemaVersion isn't the latest schema version.
-		if err != nil || neededSchemaVersion < do.InfoSchema().SchemaMetaVersion() {
-			logutil.BgLogger().Info("do not update self schema version to etcd",
-				zap.Int64("usedSchemaVersion", usedSchemaVersion),
-				zap.Int64("neededSchemaVersion", neededSchemaVersion), zap.Error(err))
-			return
-		}
+	if is := do.infoCache.GetByVersion(neededSchemaVersion); is != nil {
+		return is, true, 0, nil, nil
+	}
 
-		err = do.ddl.SchemaSyncer().UpdateSelfVersion(context.Background(), neededSchemaVersion)
-		if err != nil {
-			logutil.BgLogger().Info("update self version failed",
-				zap.Int64("usedSchemaVersion", usedSchemaVersion),
-				zap.Int64("neededSchemaVersion", neededSchemaVersion), zap.Error(err))
-		}
-	}()
+	currentSchemaVersion := int64(0)
+	if oldInfoSchema := do.infoCache.GetLatest(); oldInfoSchema != nil {
+		currentSchemaVersion = oldInfoSchema.SchemaMetaVersion()
+	}
 
+	// TODO: tryLoadSchemaDiffs has potential risks of failure. And it becomes worse in history reading cases.
+	// It is only kept because there is no alternative diff/partial loading solution.
+	// And it is only used to diff upgrading the current latest infoschema, if:
+	// 1. Not first time bootstrap loading, which needs a full load.
+	// 2. It is newer than the current one, so it will be "the current one" after this function call.
+	// 2. There are less 100 diffs.
 	startTime := time.Now()
-	ok, relatedChanges, err := do.tryLoadSchemaDiffs(m, usedSchemaVersion, neededSchemaVersion)
-	if err != nil {
+	if currentSchemaVersion != 0 && neededSchemaVersion > currentSchemaVersion && neededSchemaVersion-currentSchemaVersion < 100 {
```

Comment:
Why also refactor `isTooOldSchema`?

Reply:
It is inlined. `neededSchemaVersion > currentSchemaVersion` is added because of staleread and snapshot will load older infoschema. To keep the behavior same(do not load schemaDiffs for history cases), it must be added.